### ISSUE TITLE
Closing connection prematurely and continuing script execution in background

### ISSUE THREAD

**Issue by vonas** [docker-library/php#1113]
I am having issues with terminating the HTTP connection to which a PHP script process is bound. It works just fine when using my local [apache2/httpd PHP installation on Arch](https://wiki.archlinux.org/index.php/Apache_HTTP_Server), but using the [`php-apache`](https://hub.docker.com/_/php) Docker image won't yield the same behavior. The script just continues executing, even though I flushed the output and sent the correct headers.

Here you can find a bare minimum example: [`https://github.com/vonas/php-docker-issue.git`](https://github.com/vonas/php-docker-issue)
Instructions for running are found in the [readme](https://github.com/vonas/php-docker-issue#readme).

**Expected behavior**

The connection terminates after the fifth line **`flush();`** because of the configured headers. Opening the script in your browser shows a blank page without any content after a few milliseconds.

**Actual behavior**

The page loads for two seconds and shows the text that is echo'd in the script.

---

I looked for `php.ini` configuration values but none of them popped out as causing this issue. I also diffed my local ini with the one in the docker container, without finding any differences that involve options that look like they could cause this issue (just logging settings, extensions, etc.). I'm not really sure what's causing this so I'm posting about it here. I didn't find any issue of similar kind and googling didn't give me promising results either (maybe I'm just bad at it).

I originally tried with the image tagged as `php:7.4-apache`, but `php:8.0.1-apache` doesn't work either.

Hope someone knows what's going on. Cheers!

**Comment by tianon** [comment]
Honestly, I'm confused why this is working the way you describe on your Arch host -- from what I can tell, the intended behavior here is that the `flush()` just sends the headers and that the body is still technically fair game?  I don't think PHP (or Apache) is supposed to parse `header('Content-Length: 0')` and act on it by killing PHP.

https://stackoverflow.com/a/141026/433558 seems to confirm that this is intended behavior.  I think if you want PHP to explicitly stop, you need to `die` or `exit` (which *are* defined as stopping the current script explicitly)?

**Comment by vonas** [comment]
> from what I can tell, the intended behavior here is that the flush() just sends the headers and that the body is still technically fair game? I don't think PHP (or Apache) is supposed to parse header('Content-Length: 0') and act on it by killing PHP.

Sorry, I might not have made my goal clear enough. Just like in the StackOverflow post, my goal is to let _the script continue in the background_ while closing the connection and letting the client think that processing is done, while in fact it's not. I am not expecting Apache to kill my PHP process but instead terminate the HTTP connection to the client.

> https://stackoverflow.com/a/141026/433558 seems to confirm that this is intended behavior.

The difference between my example and that example of the person who asked the question is, that I am not trying to output any data, so afaict there's no need to handle that output buffering issue. That it's working on my local host confirms this to me.

Replacing my code with the proposed solution doesn't produce the intended and described behavior with the docker image(s) in question. On the other hand, running the exact solution on my local host works just fine.

**Comment by yosifkit** [comment]
That is a very resource intense way to handle background tasks. To limit the number of processes spawned, most PHP frameworks have a set of worker processes and a queue though which user pages add items to the worker queue.

I'd guess that `gzip compression => enabled` could be the issue (https://stackoverflow.com/a/10465055), but really you'd need to compare the `php -i` output between the two instances to see where they differ.

This seems like a configuration issue and so I don't see this as an issue with the image and recommend we close.

**Comment by vonas** [comment]
> That is a very resource intense way to handle background tasks.

You're right, it is. The endpoint I'm using this for requires authentication and is called at most once a day, so I chose this approach because it was the simplest and saved precious time for a rarely executed background task. But it's a poor design choice in general, I agree. Thanks for the perspective on that.

The compression thing goes in the right direction. I've looked online a little bit more and tried out different approaches of people in the comments of the PHP docs. I found this to work: https://www.php.net/manual/de/features.connection-handling.php#112843

With a little trial and error it turned out that this line made the difference:
```php
header("Content-Encoding: none\r\n");
```

Without it the server sends `Content-Encoding: gzip`. I then very quickly found this post that states the exact same problem that I am having: https://stackoverflow.com/questions/10267061/for-php-flush-how-to-disable-gzip-for-specific-file

Thank you both for your time! Let's close this.
